File: MigrationV3V4/CSharp/Skills/V3SimpleSandwichBot/Authentication/AllowedCallersClaimsValidator.cs
using Microsoft.Bot.Connector.SkillAuthentication;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;
using System.Web;

namespace Microsoft.Bot.Sample.SimpleSandwichBot.Authentication
{
    /// <summary>
    /// Sample claims validator that loads an allowed list from configuration if present
    /// and checks that requests are coming from allowed parent bots.
    /// </summary>
    public class AllowedCallersClaimsValidator : ClaimsValidator
    {
        private readonly IList<string> _allowedCallers;

        public AllowedCallersClaimsValidator(IList<string> allowedCallers)
        {
            // AllowedCallers is the setting in web.config file
            // that consists of the list of parent bot ids that are allowed to access the skill
            // to add a new parent bot simply go to the AllowedCallers and add
            // the parent bot's microsoft app id to the list

            _allowedCallers = allowedCallers ?? throw new ArgumentNullException(nameof(allowedCallers));
        }

        public override Task ValidateClaimsAsync(IList<Claim> claims)
        {
            // if _allowedCallers is null we allow all calls
            if (_allowedCallers != null && SkillValidation.IsSkillClaim(claims))
            {
                // Check that the appId claim in the skill request is in the list of skills configured for this bot.
                var appId = JwtTokenValidation.GetAppIdFromClaims(claims);
                if (!_allowedCallers.Contains(appId))
                {
                    throw new UnauthorizedAccessException($"Received a request from a bot with an app ID of \"{appId}\". To enable requests from this caller, add the app ID to your configuration file.");
                }
            }

            return Task.CompletedTask;
        }
    }
}

File: Migration/MigrationV3V4/Node/Skills/v3-booking-bot-skill/allowedCallersClaimsValidator.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const skills = require('botbuilder/skills-validator');
const path = require('path');

// Import required bot configuration.
const ENV_FILE = path.join(__dirname, '.env');
require('dotenv').config({ path: ENV_FILE });

// Load the AppIds for the configured callers (we will only allow responses from skills we have configured).
// process.env.AllowedCallers is the list of parent bot Ids that are allowed to access the skill
// to add a new parent bot simply go to the .env file and add
// the parent bot's Microsoft AppId to the list under AllowedCallers, e.g.:
//  AllowedCallers=195bd793-4319-4a84-a800-386770c058b2,38c74e7a-3d01-4295-8e66-43dd358920f8
const allowedCallers = [process.env.ROOT_BOT_APP_ID]; // process.env.AllowedCallers ? process.env.AllowedCallers.split(',') : undefined;

/**
 * Sample claims validator that loads an allowed list from configuration if present
 * and checks that requests are coming from allowed parent bots.
 * @param claims An array of Claims decoded from the HTTP request's auth header
 */
const allowedCallersClaimsValidator = async (claims) => {
    if (!allowedCallers || allowedCallers.length == 0) {
        throw new Error(`DefaultAuthenticationConfiguration allowedCallers must contain at least one element of '*' or valid MicrosoftAppId(s).`);
    }
    if (!claims || claims.length < 1) {
        throw new Error(`DefaultAuthenticationConfiguration.validateClaims.claims parameter must contain at least one element.`);
    }
    // If allowedCallers is undefined we allow all calls
    // If allowedCallers contains '*' we allow all callers
    if (skills.SkillValidation.isSkillClaim(claims)) {
                
        if(allowedCallers[0] === '*') {
            return;
        }
        // Check that the appId claim in the skill request is in the list of skills configured for this bot.
        const appId = skills.JwtTokenValidation.getAppIdFromClaims(claims);
        if (allowedCallers.includes(appId)) {
            return;
        }
        throw new Error(`Received a request from a bot with an app ID of "${ appId }". To enable requests from this caller, add the app ID to your configuration file.`);
    }
    throw new Error(`DefaultAuthenticationConfiguration.validateClaims called without a Skill claim in claims.`);
};

module.exports.allowedCallersClaimsValidator = allowedCallersClaimsValidator;
